src/snapshot/metadata.rs

```diff
@@ -99,11 +100,57 @@ impl Default for InstanceMetadata {
 }
 
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct InstanceContext {}
+pub struct InstanceContext {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub path_ignore_rules: Arc<Vec<PathIgnoreRule>>,
+}
+
+impl InstanceContext {
+    /// Extend the list of ignore rules in the context with the given new rules.
+    pub fn add_path_ignore_rules<I: IntoIterator<Item = PathIgnoreRule>>(&mut self, new_rules: I) {
+        let new_rules = new_rules.into_iter();
+
+        // If the iterator is empty, we can skip cloning our list of ignore
+        // rules and appending to it.
+        if let Some(upper_bound) = new_rules.size_hint().1 {
```

Comment:
It probably won't make a difference in practice, but size_hint is not required to give accurate results. Every place this is used also implements ExactSizeIterator, so you can add that trait bound and check is_empty.

Reply:
According to the docs, if `size_hint` provides a non-`None` upper bound and it's wrong, we can consider the iterator to be wrong.

Requiring `ExactSizeIterator` is a good idea though, I think it'll make this code a little clearer and catch any future misuses of this function.